shop4All-backend/src/main/java/com/example/shop4All_backend/services/ProductService.java

```diff
@@ -19,8 +26,92 @@ public class ProductService {
 
     //add a product
     public Product addNewProduct(Product product) {
+        validateCredentials(product);
         UserDetails userDetails = (UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
         product.setCompanySeller(userRepo.findByUserEmail(userDetails.getUsername()).get().getUserCompanyName());
         return productRepo.save(product);
     }
+
+    //get a product by productId
+    public Product getProductById(Integer productId) {
+        Optional<Product> optionalProduct = productRepo.findById(productId);
+        if (optionalProduct.isPresent()) {
+            return optionalProduct.get();
+        }
+        throw new ProductException("Product not found");
+    }
+
+    //returned products that are only sold by the logged seller
+    public List<Product> getAllProductsByCompany() {
+        UserDetails userDetails = (UserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        String companySeller = userRepo.findByUserEmail(userDetails.getUsername()).get().getUserCompanyName();
+        List<Product> products = (List<Product>) productRepo.findAll();
+        List<Product> filteredProducts = new ArrayList<>();
+        for (Product product : products) {
+            if (companySeller.equals(product.getCompanySeller())) {
+                filteredProducts.add(product);
+            }
+        }
+        return filteredProducts;
+    }
+
+    //delete a product from database
+    public Product deleteProduct(Integer productId) {
+        Product deletedProduct = productRepo.findById(productId).get();
+        productRepo.deleteById(productId);
+        return deletedProduct;
+    }
+
+    //add discount to a product
+    public Product setDiscount(Product product) {
+        checkDiscount(product);
+        Product discountedProduct = checkProduct(product.getProductId());
+        discountedProduct.setProductDiscounted(product.getProductDiscounted());
+        discountedProduct.setProductFromDiscounted(product.getProductFromDiscounted());
+        discountedProduct.setProductToDiscounted(product.getProductToDiscounted());
+        return productRepo.save(discountedProduct);
+    }
+
+    //delete a discount from a product
+    public Product deleteDiscount(Integer productId) {
+        Product deletedProduct = checkProduct(productId);
+        deletedProduct.setProductFromDiscounted(null);
+        deletedProduct.setProductToDiscounted(null);
+        deletedProduct.setProductDiscounted(0.0d);
+        return productRepo.save(deletedProduct);
+    }
+
+    //check the discount fields
+    public void checkDiscount(Product product) {
+        LocalDate today = LocalDate.now();
+        if (product.getProductFromDiscounted().isBefore(today)) {
+            throw new ProductException("Your start date is before today.");
```

Comment:
You can add today variable as a string in the Exception message. You can also add the product.getProductFromDiscounted as a string in the Exception message for easier debugging in the future.